rpe/resources/gcp.py

```diff
@@ -105,6 +106,49 @@ def type(self):
 
         return ".".join(type_components)
 
+    # Google's documentation describes what it calls a 'full resource name' for
+    # resources. None of the API's seem to implement it (except Cloud Asset
+    # Inventory). This attempts to generate it from the discovery-based api
+    # client's generated http request url.
+    #
+    # If we inject it into the resource, we can use it in policy evaluation to
+    # simplify the structure of our policies
+    def full_resource_name(self):
+
+        # If this is a resource property, return the resource's frn instead
+        if self.is_property():
+            return self.parent_resource.full_resource_name()
```

Comment:
The naming is odd because of how the code evolved. Resources should probably be renamed assets to match CAI. But a "resource" is either a "resource" such as a bucket, or a "resource property" such as the IAM policy for a bucket. 

So CAI calls them both assets, and you select if you want "resource"-type assets, or "iam-policy"-type assets. 

So when we want to full resource name for a buckets IAM policy, we really want the full resource name for the bucket

Reply:
I'm wondering if this code should extend your strategy to generate FRNs from the URI to resource properties like IAM policies?

The `getIamPolicy` requests have URIs that can be used for this, though there's a few conventions.  Some are treated like subresources:
```
>>> buckets.getIamPolicy(bucket='wooo').uri
'https://www.googleapis.com/storage/v1/b/wooo/iam?alt=json'
```
In other cases, it's a collection + RPC name:
```
>>> pubsub.getIamPolicy(resource='projects/wooo/topics/wooo').uri
https://pubsub.googleapis.com/v1/projects/wooo/topics/wooo:getIamPolicy?alt=json
>>> projects.getIamPolicy(resource='projects/wooo').uri
https://cloudresourcemanager.googleapis.com/v1/projects/projects%2Fwooo:getIamPolicy?alt=json
```

The FRN doc covers both of these cases, even though the second case produces weird looking FRNs.

Reply:
I see what you mean, in the bucket example /buckets/<bucket_name>/iam could reasonably be considered the full resource name for a bucket's IAM policy.  I'm concerned we'd run into a bunch of corner cases going that route, and you've already mentioned 2.

The CAI solution is to build a structure that looks like this:

```
{
  "name": "//storage.googleapis.com/cleardata-nonprod-playground.appspot.com",
  "asset_type": "storage.googleapis.com/Bucket",
  "iam_policy": {...},
  "resource": {...}
}
```

However, you can only request the IAM policy *OR* the resource with the exportAssets call (You need to do 2 exports to get both, so you never see both the _iam-policy_ key and the _resource_ key in the same object.)

I think its best to follow what they came up with (or as close as possible).

Reply:
yes, agreed - looks good